ReactiveCocoa/Swift/SignalProducer.swift

```diff
@@ -1745,75 +1730,42 @@ extension SignalProducer {
 					}
 				}
 			}
-		}
 
-		let bufferingObserver: Signal<Value, Error>.Observer = Observer { event in
-			let originalState = state.modify { state in
-				if let value = event.value {
-					state.add(value, upTo: capacity)
-				} else {
-					// Disconnect all observers and prevent future
-					// attachments.
-					state.terminationEvent = event
-					state.observers = nil
-				}
-			}
-
-			originalState.observers?.forEach { $0.action(event) }
+			// Start the underlying producer if it has never been started.
+			bootstrap.dispose()
 		}
+	}
+}
+
+private final class DeallocationToken {
+	let (deallocSignal, observer) = Signal<(), NoError>.pipe()
 
-		return (producer, bufferingObserver)
+	deinit {
+		observer.sendCompleted()
 	}
 }
 
-/// A uniquely identifying token for Observers that are replaying values in
-/// BufferState.
+/// A uniquely identifying token for `Observer`s that are replaying values in
+/// `ReplayState`.
 private final class ReplayBuffer<Value> {
 	private var values: [Value] = []
 }
 
-private struct BufferState<Value, Error: ErrorProtocol> {
-	/// All values in the buffer.
+private struct ReplayState<Value, Error: ErrorProtocol> {
+	/// All cached values.
 	var values: [Value] = []
 
-	/// Any terminating event sent to the buffer.
+	/// A terminating event emitted by the underlying producer.
 	///
 	/// This will be nil if termination has not occurred.
 	var terminationEvent: Event<Value, Error>?
 
-	/// The observers currently attached to the buffered producer, or nil if the
-	/// producer was terminated.
+	/// The observers currently attached to the caching producer, or `nil` if the
+	/// caching producer was terminated.
 	var observers: Bag<Signal<Value, Error>.Observer>? = Bag()
 
 	/// The set of unused replay token identifiers.
 	var replayBuffers: Bag<ReplayBuffer<Value>> = Bag()
-
-	/// Appends a new value to the buffer, trimming it down to the given capacity
-	/// if necessary.
-	mutating func add(_ value: Value, upTo capacity: Int) {
-		precondition(capacity >= 0)
-
-		for buffer in replayBuffers {
-			buffer.values.append(value)
-		}
-
-		if capacity == 0 {
-			values = []
-			return
-		}
-
-		if capacity == 1 {
-			values = [ value ]
-			return
-		}
-
-		values.append(value)
-
-		let overflow = values.count - capacity
-		if overflow > 0 {
-			values.removeSubrange(0..<overflow)
-		}
-	}
```

Comment:
This seems like it's still useful to have as a `mutating func`. Keeping the state manipulation here makes it a little easier to follow the higher-level threading logic in the producer.